File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/supabase-table-missing-rls.ts
import { defineRule } from "../../utils/define-rule.js";
import type { ScanFinding } from "../../utils/file-scan.js";
import { escapeRegExp } from "./utils/escape-reg-exp.js";
import { getLocationAtIndex } from "./utils/get-location-at-index.js";
import { isSupabaseMigrationPath } from "./utils/is-supabase-migration-path.js";

// A `create table` for a public-schema table — the only schema PostgREST
// exposes to the anon key. Unqualified names default to `public`, so they
// count; internal/Supabase-managed schemas (`auth.`, `storage.`, a `private.`
// schema, …) are skipped via the negative lookahead. Requiring `(` or `as`
// after the name keeps `-- create table …` SQL comments and prose out of the
// match. Group 1 captures the table name for the per-table RLS check.
const CREATE_PUBLIC_TABLE_PATTERN =
  /create\s+(?:unlogged\s+)?table\s+(?:if\s+not\s+exists\s+)?(?!(?:auth|storage|realtime|vault|extensions|graphql|graphql_public|pgbouncer|net|supabase_functions|supabase_migrations|cron|pgsodium|pgmq|information_schema|pg_catalog|pg_temp|private|internal)\s*\.)(?:public\s*\.\s*)?["`]?([A-Za-z_][\w$]*)["`]?(?:\s*\(|\s+as\b)/gi;

// Only `alter table <name> enable row level security` makes a public table
// safe. A `create policy` alone does NOT — policies are inert until RLS is
// enabled — so RLS must be checked per table rather than file-wide (a sibling
// table enabling RLS must not vouch for this one). The enable keyword must
// follow the table name directly so a nearby unrelated `enable` cannot match.
const enableRlsForTablePattern = (tableName: string): RegExp =>
  new RegExp(
    `alter\\s+table\\s+(?:if\\s+exists\\s+)?(?:only\\s+)?(?:public\\s*\\.\\s*)?["\`]?${escapeRegExp(tableName)}["\`]?\\s+(?:force\\s+)?enable\\s+row\\s+level\\s+security`,
    "i",
  );

export const supabaseTableMissingRls = defineRule({
  id: "supabase-table-missing-rls",
  title: "Supabase table created without Row Level Security",
  severity: "error",
  recommendation:
    "Enable RLS in the same migration (`alter table <name> enable row level security;`) and add `auth.uid()`-scoped policies for select/insert/update/delete. A public table without RLS is fully readable and writable with the public anon key.",
  scan: (file) => {
    if (!isSupabaseMigrationPath(file.relativePath)) return [];
    const content = file.content;
    if (!/create\s+(?:unlogged\s+)?table/i.test(content)) return [];

    const findings: ScanFinding[] = [];
    CREATE_PUBLIC_TABLE_PATTERN.lastIndex = 0;
    for (
      let match = CREATE_PUBLIC_TABLE_PATTERN.exec(content);
      match !== null;
      match = CREATE_PUBLIC_TABLE_PATTERN.exec(content)
    ) {
      const tableName = match[1];
      if (tableName === undefined) continue;
      if (enableRlsForTablePattern(tableName).test(content)) continue;
      const location = getLocationAtIndex(content, match.index);
      findings.push({
        message:
          "Supabase migration creates a public table but never enables Row Level Security, leaving every row exposed to the anon key.",
        line: location.line,
        column: location.column,
      });
    }
    return findings;
  },
});

File: native/oxlint/scans/supabase-table-missing-rls.rs
use lazy_regex::{Lazy, Regex, lazy_regex};
use rustc_hash::FxHashMap;

use super::{
    ScanFinding, get_location_at_index::get_location_at_index,
    sanitize_sql_for_scan::sanitize_sql_for_scan,
};

const MESSAGE: &str = "Supabase migration creates a public table but never enables Row Level Security, leaving every row exposed to the anon key.";

static CREATE_TABLE_PREFILTER_PATTERN: Lazy<Regex> =
    lazy_regex!(r"(?i)create\s+(?:unlogged\s+)?table");
static CREATE_TABLE_PATTERN: Lazy<Regex> = lazy_regex!(
    r#"(?i)create\s+(?:unlogged\s+)?table\s+(?:if\s+not\s+exists\s+)?(?:(["`]?[A-Za-z_][A-Za-z0-9_$]*["`]?)\s*\.\s*)?(["`]?[A-Za-z_][A-Za-z0-9_$]*["`]?)\s*(?:\(|as(?-u:\b))"#
);
static ENABLE_RLS_PATTERN: Lazy<Regex> = lazy_regex!(
    r#"(?i)alter\s+table\s+(?:if\s+exists\s+)?(?:only\s+)?(?:(["`]?[A-Za-z_][A-Za-z0-9_$]*["`]?)\s*\.\s*)?(["`]?[A-Za-z_][A-Za-z0-9_$]*["`]?)\s+(?:force\s+)?enable\s+row\s+level\s+security"#
);
static ENABLE_RLS_KEYWORD_PATTERN: Lazy<Regex> =
    lazy_regex!(r"(?i-u:\benable\s+row\s+level\s+security\b)");

pub fn scan(relative_path: &str, source: &str) -> Vec<ScanFinding> {
    if !is_supabase_migration_path(relative_path) {
        return Vec::new();
    }
    let sanitized = sanitize_sql_for_scan(source);
    if !CREATE_TABLE_PREFILTER_PATTERN.is_match(&sanitized) {
        return Vec::new();
    }

    let mut last_enable_index_by_table = FxHashMap::default();
    let mut static_enable_count = 0;
    for captures in ENABLE_RLS_PATTERN.captures_iter(&sanitized) {
        let Some(full_match) = captures.get(0) else {
            continue;
        };
        let schema = captures.get(1).map(|capture| sql_name(capture.as_str()));
        if schema.as_deref().is_some_and(|name| name != "public") {
            continue;
        }
        let Some(table) = captures.get(2) else {
            continue;
        };
        static_enable_count += 1;
        last_enable_index_by_table.insert(sql_name(table.as_str()), full_match.start());
    }
    if ENABLE_RLS_KEYWORD_PATTERN.find_iter(&sanitized).count() > static_enable_count {
        return Vec::new();
    }

    let mut findings = Vec::new();
    for captures in CREATE_TABLE_PATTERN.captures_iter(&sanitized) {
        let Some(full_match) = captures.get(0) else {
            continue;
        };
        let schema = captures.get(1).map(|capture| sql_name(capture.as_str()));
        if schema.as_deref().is_some_and(|name| name != "public") {
            continue;
        }
        let Some(table) = captures.get(2) else {
            continue;
        };
        if last_enable_index_by_table
            .get(&sql_name(table.as_str()))
            .is_some_and(|enable_index| *enable_index >= full_match.start())
        {
            continue;
        }
        let (line, column) = get_location_at_index(source, &sanitized, full_match.start());
        findings.push(ScanFinding::inherited(MESSAGE, line, column));
    }
    findings
}

fn is_supabase_migration_path(relative_path: &str) -> bool {
    ["supabase/migrations/", "supabase/schemas/"].iter().any(|segment| {
        relative_path.starts_with(segment) || relative_path.contains(&format!("/{segment}"))
    })
}

fn sql_name(name: &str) -> String {
    name.trim_matches(['"', '`']).to_ascii_lowercase()
}
